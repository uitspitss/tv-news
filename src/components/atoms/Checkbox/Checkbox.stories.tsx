import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import base from 'paths.macro';

import { Checkbox, CheckboxProps } from './Checkbox';

const path = base.replace(
  /^.+\/(atoms|molecules|organisms|templates|pages)\/(.+)\//,
  '$1/$2',
);

export default {
  title: `${path}`,
  component: Checkbox,
} as Meta;

const Template: Story<CheckboxProps> = (args) => <Checkbox {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'label',
};
