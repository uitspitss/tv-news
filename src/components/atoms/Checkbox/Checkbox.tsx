import React, {
  memo,
  ComponentPropsWithRef,
  ForwardRefExoticComponent,
  forwardRef,
} from 'react';
import tw, { styled } from 'twin.macro';

export type CheckboxProps = {
  /**
   * class for twin.macro
   */
  className?: string;
  /**
   * ラベル
   */
  label: string;
  /**
   * error message
   */
  error?: string;
} & ComponentPropsWithRef<'input'>;

type Props = CheckboxProps;

const Component: ForwardRefExoticComponent<Props> = forwardRef((props, ref) => {
  const { className, name, label, error } = props;

  return (
    <div className={className}>
      {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
      <label htmlFor={name}>
        <input
          name={name}
          type="checkbox"
          id={name}
          aria-label={label}
          aria-invalid={error ? 'true' : 'false'}
          ref={ref}
        />
        <span className="label">{label}</span>
      </label>
      {error && (
        <div>
          <span className="error" role="alert">
            {error}
          </span>
        </div>
      )}
    </div>
  );
});

const StyledComponent = styled(Component)`
  & > {
    ${tw`flex items-center`}

    label > input {
      ${tw`mr-2 rounded`}
    }

    div > .error {
      ${tw`text-red-500`}
    }
  }
`;

/**
 * Checkbox
 */
export const Checkbox: ForwardRefExoticComponent<CheckboxProps> = memo(
  forwardRef((props, ref) => {
    return <StyledComponent {...props} ref={ref} />;
  }),
);
