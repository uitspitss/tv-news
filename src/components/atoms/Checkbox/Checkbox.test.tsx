import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import { Checkbox } from './Checkbox';

describe('atoms/Checkbox', () => {
  test('snapshot test', () => {
    const { container } = render(<Checkbox label="checkbox label" />);
    expect(container).toMatchSnapshot();
  });

  describe('Behavior', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      render(<Checkbox label="checkbox label" />);
    });

    test('initial view', () => {
      expect(screen.getByLabelText('checkbox label')).toBeInTheDocument();
      expect(screen.getByText('checkbox label')).toBeInTheDocument();
    });

    test('click, then checked', () => {
      expect(screen.getByLabelText('checkbox label')).not.toBeChecked();
      fireEvent.click(screen.getByLabelText('checkbox label'));
      expect(screen.getByLabelText('checkbox label')).toBeChecked();
    });
  });
});
