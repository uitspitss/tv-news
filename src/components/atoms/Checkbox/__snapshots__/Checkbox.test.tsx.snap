// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/Checkbox snapshot test 1`] = `
.emotion-0> {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0>label>input {
  margin-right: 0.5rem;
  border-radius: 0.25rem;
}

.emotion-0>div>.error {
  --tw-text-opacity: 1;
  color: rgba(239, 68, 68, var(--tw-text-opacity));
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <label>
      <input
        aria-invalid="false"
        aria-label="checkbox label"
        type="checkbox"
      />
      <span
        class="label"
      >
        checkbox label
      </span>
    </label>
  </div>
</div>
`;
